### ISSUE TITLE
PROVISIONING_PROFILE_SPECIFIER and PROVISIONING_PROFILE changes in Xcode 8.3 causing white labelled app to fail to sign

### ISSUE THREAD

**Issue by jdouglas-nz** [fastlane/fastlane#8973]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)

### Issue Description
Since the migration from XCode 8.2.1 -> 8.3, our white labelled app (which we parameterize app identifier and app name among other things) now fails to sign. 

The relevant piece of our _Fastfile_ is this: 

``` 
        # ensure there is a private key and signing cert available for the app.
        match(
            app_identifier: newAppId,
            git_url: 'ssh://some/path/to/git/repo.git',
            type: 'appstore',
            username: '[redacted]'
        )

        # ensure there is a provisioning profile available to build and sign the app with
        sigh(
            app_identifier: newAppId,
            provisioning_name: "match AppStore #{newAppId}"
        )

        # update app identifier so can sign with newly obtained provisioning profile
        update_app_identifier(
            xcodeproj: Paths::PROJECT_PATH,
            plist_path: Paths::PLIST_FILE_PATH,
            app_identifier: newAppId
        )

        # set the project provisoning profile UUID to the mobile provisioning profile obtained from sigh
        update_project_provisioning(
            xcodeproj: Paths::PROJECT_PATH,
            target_filter: '.*',
            profile: "./AppStore_#{newAppId}.mobileprovision"
        )

        # finally, build app
        gym(
            scheme: Schemes::APP_SCHEME,
            workspace: Paths::WORKSPACE_PATH
        )
```

after getting to the end of gym, and starting to sign, there are a bunch of these spat out:

> ❌  Code signing is required for product type 'Application' in SDK 'iOS 10.3'

XCode 8.3 and Apple's related build tooling now seems to use **BOTH** the `PROVISIONING_PROFILE_SPECIFIER` and `PROVISIONING_PROFILE` **TOGETHER** to determine if a particular provisioning profile can be used. 

We have confirmed this by after running the script up until `gym`, we have manually changed the `PROVISIONING_PROFILE_SPECIFIER` as well. After this change, code signing works as it did. 

Because not everyone white labels their apps, may I suggest the way to 'solve' this problem is to add a profile_specifier flag to the [update_project_code_signing](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_project_code_signing.rb) action? it doesn't seem that difficult, and I am happy to perform the change if this is the agreed approach.

For now, to unblock our builds, we are just regexing our 'white-labelled' specifier with an injected replacement.
 
# Environment
<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.12                                       |
| Ruby                        | 2.2.4                                       |
| Bundler?                    | false                                       |
| Git                         | git version 2.11.0 (Apple Git-81)           |
| Installation Source         | ~/.fastlane/bin/bundle/bin/fastlane         |
| Host                        | Mac OS X 10.12 (16A323)                     |
| Ruby Lib Dir                | ~/.fastlane/bin/bundle/lib                  |
| OpenSSL Version             | OpenSSL 1.0.2g  1 Mar 2016                  |
| Is contained                | true                                        |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 8.3                                         |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE | en_US.UTF-8 | ✅ |

### fastlane files:

**No Fastfile found**

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.28.3  | ✅ Up-To-Date  |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| CFPropertyList            | 2.3.5        |
| activesupport             | 4.2.7.1      |
| addressable               | 2.5.0        |
| babosa                    | 1.0.2        |
| bigdecimal                | 1.2.6        |
| bundler                   | 1.14.3       |
| claide                    | 1.0.1        |
| colored                   | 1.2          |
| colored2                  | 3.1.2        |
| commander                 | 4.4.3        |
| commander-fastlane        | 4.4.4        |
| domain_name               | 0.5.20161129 |
| dotenv                    | 2.2.0        |
| excon                     | 0.54.0       |
| faraday                   | 0.11.0       |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.11.0.1     |
| fastimage                 | 2.0.1        |
| gh_inspector              | 1.0.3        |
| google-api-client         | 0.9.26       |
| googleauth                | 0.5.1        |
| highline                  | 1.7.8        |
| http-cookie               | 1.0.3        |
| httpclient                | 2.8.3        |
| hurley                    | 0.2          |
| i18n                      | 0.7.0        |
| io-console                | 0.4.3        |
| json                      | 1.8.1        |
| jwt                       | 1.5.6        |
| libxml-ruby               | 2.9.0        |
| little-plugger            | 1.1.4        |
| logging                   | 2.1.0        |
| memoist                   | 0.15.0       |
| mime-types                | 3.1          |
| mime-types-data           | 3.2016.0521  |
| mini_magick               | 4.5.1        |
| minitest                  | 5.4.3        |
| multi_json                | 1.12.1       |
| multi_xml                 | 0.6.0        |
| multipart-post            | 2.0.0        |
| nanaimo                   | 0.2.3        |
| nokogiri                  | 1.5.6        |
| os                        | 0.9.6        |
| plist                     | 3.2.0        |
| power_assert              | 0.2.2        |
| psych                     | 2.0.8        |
| public_suffix             | 2.0.5        |
| rake                      | 10.4.2       |
| rdoc                      | 4.2.0        |
| representable             | 2.3.0        |
| retriable                 | 2.1.0        |
| rouge                     | 1.11.1       |
| rubygems-update           | 2.5.2        |
| rubyzip                   | 1.2.0        |
| security                  | 0.1.3        |
| signet                    | 0.7.3        |
| slack-notifier            | 1.5.1        |
| sqlite3                   | 1.3.13       |
| terminal-notifier         | 1.7.1        |
| terminal-table            | 1.7.3        |
| test-unit                 | 3.0.8        |
| thread_safe               | 0.3.5        |
| tty-screen                | 0.5.0        |
| tzinfo                    | 1.2.2        |
| uber                      | 0.0.15       |
| unf                       | 0.1.4        |
| unf_ext                   | 0.0.7.2      |
| unicode-display_width     | 1.1.3        |
| word_wrap                 | 1.0.0        |
| xcodeproj                 | 1.4.4        |
| xcpretty                  | 0.2.4        |
| xcpretty-travis-formatter | 0.0.4        |
</details>

*generated on:* **2017-04-24**
</details>

Thanks!

**Comment by glesage** [comment]
Been having the same issue https://github.com/fastlane/fastlane/issues/8821#issuecomment-292876401

**Comment by jdouglas-nz** [comment]
@glesage, thanks for the ref! I did have a look at your ticket before creating one, and it didn't seem that @SiberianCMS was discussing about the `ProvisioningStyle = Manual;` stuff was related. 


Only if you update both the `PROVISIONING_PROFILE` and `PROVISIONING_PROFILE_SPECIFIER` to the profile of the newly updated `PRODUCT_BUNDLE_IDENTIFIER` which is altered by running `update_app_identifier` as per the above snippet of a `Fastfile`

**Comment by glesage** [comment]
I see, I think what you posted is actually a step further than I was at the time (:

I'll have to do the regex approach as well for now

**Comment by brockklein** [comment]
@jdouglas-nz @glesage I'm a bit of an amateur using Fastlane within a white label context as well. Any chance either of you would be willing to share your final work around? 

I'm not a regex pro and would love to unabashedly stand on the shoulders of giants to get around this issue.

**Comment by brockklein** [comment]
@jdouglas-nz @glesage nix that request. I was able to get everything working with:

```
match(app_identifier: @bundle_id, type: "appstore", username: @user)

    # ensure there is a provisioning profile available to build and sign the app with
    sigh(
      app_identifier: @bundle_id,
      provisioning_name: "match AppStore #{@bundle_id}",
      username: @user
    )

    # set the project provisoning profile UUID to the mobile provisioning profile obtained from sigh
    update_project_provisioning(
        xcodeproj: "EventCast.xcodeproj",
        target_filter: '.*',
        profile: "./AppStore_#{@bundle_id}.mobileprovision"
    )
```

(variable names my own of course)

**Comment by col** [comment]
I'm having the same issue as this and, at least in my case, it cannot be resolved by using `sigh` as I do not have direct access to the ADC account.

Are there any other ways to set the 'PROVISIONING_PROFILE_SPECIFIER' in the .xcodeproj correctly?

**Comment by col** [comment]
I'm pretty sure the `update_project_provisioning` action needs to be updated to set the `PROVISIONING_PROFILE_SPECIFIER` as well as `PROVISIONING_PROFILE`. 

See: [update_project_provisioning.rb](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/update_project_provisioning.rb) Line 66

Can someone else validate that this is the right approach? I can put a PR together but what to confirm this is the right approach.

**Comment by col** [comment]
Ok, I've found a workaround but to be honest I'm pretty surprised how difficult it is to build an app with a different set of signing details. Here's what I have now that "appears" to be working. (The real test will be tomorrow when I _actually_ upload to the AppStore).

```
desc "Create IPA for AppStore submission"
lane :prod_release do    

  update_app_identifier(
    plist_path: "TestApp/Info.plist",
    app_identifier: "com.example.appname"
  )    

  update_project_team(
    path: "TestApp.xcodeproj",
    teamid: "ABCDE12345"
  )

  update_project_provisioning(
    xcodeproj: "TestApp.xcodeproj",
    profile: "~/Library/MobileDevice/Provisioning\ Profiles/007e69e4-1234-5678-9012-3ch008c95da1.mobileprovision",
    target_filter: "TestApp",
    build_configuration: "Release"
  )

  gym(
    codesigning_identity: "iPhone Distribution: Example (ACBDE12345)",      
    export_method: "app-store",
    export_options: {
      provisioningProfiles: { 
        "com.example.appname" => "AppStore com.example.appname"
      }
    },
    xcargs: "PROVISIONING_PROFILE_SPECIFIER='AppStore com.example.appname'"
  )
end
```

I definitely think there is room here for a specific fastlane action that handles all of this mess. Something like:

```
desc "Create IPA for AppStore submission"
lane :prod_release do    

  update_code_signing_options(
    path: "TestApp.xcodeproj",
    app_identifier: "com.example.appname",
    team_id: "ABCDE12345",
    provisioning_profile: "AppStore com.example.appname",    
  )    

  gym(
    codesigning_identity: "iPhone Distribution: Example (ACBDE12345)",      
    export_method: "app-store",
  )
end
```

Or am I just over simplifying this?
